### DISCUSSION TITLE
Resolve Audience: Different Audience in ID Token and Access Token

### DISCUSSION THREAD

**Discussion by Zacherl** [keycloak/keycloak#14398] [Q&A] [answered]
Hi all,

I am new to Keycloak and loving most of it. However, I am currently stuck with validating the audience.

I created a new client "client_xy" and created some client roles. The roles are stored properly in the ID and access tokens at "resource_access.client_xy.roles", however, the audience claim "aud" is behaving strangely: In the ID token the client ID is used ("aud": "client_xy"), while in the access token the account client ID is used ("aud": "account"). Is there some obvious error in the configuration that could cause that behavior? I also need "aud": "client_xy" in the access token, not only in the ID token.

I am on version 19.0.2.

Any advice would be much appreciated!

**Comment by sschu** [comment]
@Zacherl That's exactly working as intended. The audience of an ID token is the client who is actually retrieving the token to get information on the user. The audience of the access token is some resource server (api) the client will pass the token to when making a request. The access token could be opaque as per OAuth2 spec so it is in general not even readable by the client.

**Comment by bonnm** [comment]
Older versions of keycloak did not add an aud claim to the access token. You can control the aud claim value by adding an "Audience" mapper to the client's (or scope's) mappers and enabling the "add to access token" switch of the mapper.

**Comment by Zacherl** [comment]
Thanks a lot you for your quick replies!

@sschu The client that retrieved the token was "client_xy", not "account" - I'm actually not using the "account" client at all. Is it still working as intended if the "aud" claim contains "account"? I expected "client_xy" in there. I am a bit confused - I even checked the "token" request containing the three tokens in its response (ID token, access token and refresh token). The ID token has "client_xy" as audience (as expected), the access token "account".

Adding an audience mapper would work, but then I would have to add an audience mapper for every client.

**Reply by sschu** [reply]
@Zacherl The access token contains all clients the user has roles of in audience if the client that retrieved the token has "Full Scope Allowed" enabled. Also see my below answer.

**Comment by bonnm** [comment, marked as answer]
If the access token actually contains the client ID of a completely uninvolved client (here: "account"), this looks like a misconfiguration or a bug.

The OAuth use case of the access token is not primarily that the frontend client (which triggered the authentication, which is in the azp claim) can check the user identity (that is what the ID token is for!). Rather, the purpose of the access token is to use it when calling a 3rd party API. This 3rd party then checks whether it itself is entered as an addressee in the aud claim.

So you really only need the audience mapper for clients that use the access token in such a 3rd party delgation scenario.

**Reply by sschu** [reply]
@bonnm The user has roles for the account endpoint, and since the client most likely has "Full Scope Allowed" enabled, the roles are included in the token and also the account client becomes part of the audience (because the token can be used at the account API). If you only want roles of certain clients in the token/certain clients in the audience, just disable "Full Scope Allowed" and select the clients and roles individually.

**Reply by bonnm** [reply]
OK, thanks for clarification. I forgot to think about the "Full Scope Allowed" flag and the "audience resolve" mappers, because I normally disable them (to keep access tokens small)...

**Reply by Zacherl** [reply]
Thank you for the input! All clients we configure will be used a third parties (different APIs) that should validate the audience.

I could not find the "Full Scope Allowed" setting in the client settings, do you mean "Direct access grants"? We probably do not need the other roles, so I would disable this setting.

Still, should the audiences in the ID token and the access token not be the same?

**Reply by sschu** [reply]
<img width="1139" alt="Screenshot 2022-09-15 at 11 01 31" src="https://user-images.githubusercontent.com/3833769/190362423-a6ac4909-686d-4ca7-b8e2-d6e8847dc25e.png">
The setting is under the "Scope" tab. And as I said, the audience of access tokens are resource servers/APIs, the audience of id tokens are clients/relying parties. That's how it is designed.

You can read here about the audience claim for ID Tokens: https://openid.net/specs/openid-connect-basic-1_0.html#IDTokenValidation
and here about the audience value for JWT access tokens:
https://datatracker.ietf.org/doc/html/rfc9068#section-4

**Reply by Zacherl** [reply]
@sschu : Thanks again! Somehow our Keycloak user interface looks completely different:

![Bildschirmfoto 2022-09-15 um 13 03 41](https://user-images.githubusercontent.com/1238091/190388512-d7ba9b94-7e62-404a-a872-3dbe37cc171b.png)

Is it a different version?

**Reply by sschu** [reply]
I am using the old admin console. I can indeed not find the setting in the new admin console. To access the old console, you have to change the administration console theme to `keycloak`from `keycloak.v2`in the realm settings. This way you can at least change the configuration and then switch back to the new console. In the meantime, I will try to find out wether this setting was just forgotten in the new console or whether it was really supposed to go away...

**Reply by Zacherl** [reply]
Thanks a lot for those tips! I could disable it (in case someone stumbles on the same thing: of course I had to change the theme for the master realm, not my specific realm).

When I disable "Full Scope Allowed", the identity token still has "aud": "client_xy". For the access token, the "aud" entry is missing.

If I configure a "audience" mapper with an "included custom audience" (see below), the "aud": "client_xy" appears. This solves my original issue - thanks again for all your input, I don't know how I could have solved that alone, you helped me a lot!

![image](https://user-images.githubusercontent.com/1238091/190404250-4350afb9-61fa-4178-92ae-4523f1052782.png)

One more note, though: If I do not use the "included custom audience", but the "included client audience" and select the same client I am currently configuring, **the audience does not appear, but is still missing in the access token.** It only works with "included custom audience".

I am still wondering if there is a bug within Keycloak and if the "client_xy" should appear automatically in the access token, without any configuration. I still think the "account" client should not be in the audience since I do not use that client for acquiring the token (or at least the "client_xy" should also be in the audience). Should I open a bug ticket?

**Reply by sschu** [reply]
The way Keycloak does it is exactly as the spec says. If your client looks into the access token, you are most likely doing something wrong.

**Reply by Zacherl** [reply]
Just to clarify, since I am really curious to find out where I am wrong:
1. A web page authenticates a user with Keycloak using Authorization Code with client "client_xy".
2. The web page uses the access token as a bearer to call an API endpoint.
3. The API validates the audience in the access token (expects "client_xy", and fails if the audience is something else, in this case it was "account").

Something in this process is wrong? I should expect "account" in the audience?

**Reply by bonnm** [reply]
The API in your scenario has to validate expecting "API" in the audience. Because the access token's aud is indicating for whom the token is made for. In your case: API. If you have such a 3rd party scenario, give the person roles in the 3rd party client and use keycloaks autmatism, or configure it manually, as describes above.

To complete this:  The id token is made for the initially autrhenticating client, so its aud is "client_xy". The refresh token is made for the OP, so the refresh's aud is the url of your keycloak realm (the issuer).

**Reply by sschu** [reply]
What @bonnm wrote...

**Reply by sschu** [reply]
To extend on that by example: the account API is an API of Keycloak where the access token can be used. For example if the user has the role `account.manage-account`, the account API will allow the user to update his account data. But the account API first has to check whether it is actually an intended audience of that token.

**Reply by Zacherl** [reply]
Thanks a lot for your explanations!

**Reply by ceng-p** [reply]
Sorry for bringing up an old thread here, but I am in the same scenario. Is the reason the authenticating client isn't shown in the aud in the access token because the client and audience have to differ?

**Reply by sschu** [reply]
The reason is that the audience of an access token is the resource server, not the client.
